In [33]:
import pandas as pd

In [34]:
file_path = 'DatosPS.csv'
df = pd.read_csv(file_path)
df

Rented Bike Count  Hour  Temperature(C)  Humidity(%)  Wind speed (m/s)  \
0                   254     0            -5.2           37               2.2   
1                   204     1            -5.5           38               0.8   
2                   173     2            -6.0           39               1.0   
3                   107     3            -6.2           40               0.9   
4                    78     4            -6.0           36               2.3   
...                 ...   ...             ...          ...               ...   
8755               1003    19             4.2           34               2.6   
8756                764    20             3.4           37               2.3   
8757                694    21             2.6           39               0.3   
8758                712    22             2.1           41               1.0   
8759                584    23             1.9           43               1.3   

      Visibility (10m)  Dew point temperature(C)  Solar Radiation (MJ/m2)  \
0                 2000                     -17.6                      0.0   
1                 2000                     -17.6                      0.0   
2                 2000                     -17.7                      0.0   
3                 2000                     -17.6                      0.0   
4                 2000                     -18.6                      0.0   
...                ...                       ...                      ...   
8755              1894                     -10.3                      0.0   
8756              2000                      -9.9                      0.0   
8757              1968                      -9.9                      0.0   
8758              1859                      -9.8                      0.0   
8759              1909                      -9.3                      0.0   

      Rainfall(mm)  Snowfall (cm) Seasons     Holiday Functioning Day  \
0              0.0            0.0  Winter  No Holiday             Yes   
1              0.0            0.0  Winter  No Holiday             Yes   
2              0.0            0.0  Winter  No Holiday             Yes   
3              0.0            0.0  Winter  No Holiday             Yes   
4              0.0            0.0  Winter  No Holiday             Yes   
...            ...            ...     ...         ...             ...   
8755           0.0            0.0  Autumn  No Holiday             Yes   
8756           0.0            0.0  Autumn  No Holiday             Yes   
8757           0.0            0.0  Autumn  No Holiday             Yes   
8758           0.0            0.0  Autumn  No Holiday             Yes   
8759           0.0            0.0  Autumn  No Holiday             Yes   

     Dia Semana        Mes Momento Dia  
0       viernes  diciembre   Madrugada  
1       viernes  diciembre   Madrugada  
2       viernes  diciembre   Madrugada  
3       viernes  diciembre   Madrugada  
4       viernes  diciembre   Madrugada  
...         ...        ...         ...  
8755    viernes  noviembre       Noche  
8756    viernes  noviembre       Noche  
8757    viernes  noviembre       Noche  
8758    viernes  noviembre       Noche  
8759    viernes  noviembre       Noche  

[8760 rows x 16 columns]

In [35]:
df.shape


(8760, 16)

In [36]:
df.head()

Rented Bike Count  Hour  Temperature(C)  Humidity(%)  Wind speed (m/s)  \
0                254     0            -5.2           37               2.2   
1                204     1            -5.5           38               0.8   
2                173     2            -6.0           39               1.0   
3                107     3            -6.2           40               0.9   
4                 78     4            -6.0           36               2.3   

   Visibility (10m)  Dew point temperature(C)  Solar Radiation (MJ/m2)  \
0              2000                     -17.6                      0.0   
1              2000                     -17.6                      0.0   
2              2000                     -17.7                      0.0   
3              2000                     -17.6                      0.0   
4              2000                     -18.6                      0.0   

   Rainfall(mm)  Snowfall (cm) Seasons     Holiday Functioning Day Dia Semana  \
0           0.0            0.0  Winter  No Holiday             Yes    viernes   
1           0.0            0.0  Winter  No Holiday             Yes    viernes   
2           0.0            0.0  Winter  No Holiday             Yes    viernes   
3           0.0            0.0  Winter  No Holiday             Yes    viernes   
4           0.0            0.0  Winter  No Holiday             Yes    viernes   

         Mes Momento Dia  
0  diciembre   Madrugada  
1  diciembre   Madrugada  
2  diciembre   Madrugada  
3  diciembre   Madrugada  
4  diciembre   Madrugada

In [37]:
df.describe()

Rented Bike Count         Hour  Temperature(C)  Humidity(%)  \
count        8760.000000  8760.000000     8760.000000  8760.000000   
mean          704.602055    11.500000       12.882922    58.226256   
std           644.997468     6.922582       11.944825    20.362413   
min             0.000000     0.000000      -17.800000     0.000000   
25%           191.000000     5.750000        3.500000    42.000000   
50%           504.500000    11.500000       13.700000    57.000000   
75%          1065.250000    17.250000       22.500000    74.000000   
max          3556.000000    23.000000       39.400000    98.000000   

       Wind speed (m/s)  Visibility (10m)  Dew point temperature(C)  \
count       8760.000000       8760.000000               8760.000000   
mean           1.724909       1436.825799                  4.073813   
std            1.036300        608.298712                 13.060369   
min            0.000000         27.000000                -30.600000   
25%            0.900000        940.000000                 -4.700000   
50%            1.500000       1698.000000                  5.100000   
75%            2.300000       2000.000000                 14.800000   
max            7.400000       2000.000000                 27.200000   

       Solar Radiation (MJ/m2)  Rainfall(mm)  Snowfall (cm)  
count              8760.000000   8760.000000    8760.000000  
mean                  0.569111      0.148687       0.075068  
std                   0.868746      1.128193       0.436746  
min                   0.000000      0.000000       0.000000  
25%                   0.000000      0.000000       0.000000  
50%                   0.010000      0.000000       0.000000  
75%                   0.930000      0.000000       0.000000  
max                   3.520000     35.000000       8.800000

In [38]:
duplicates = len(df[df.duplicated()])
print(f'Number of Duplicate Entries: {duplicates}')

Number of Duplicate Entries: 0


In [39]:
# número de valores perdidos
missing_values = df.isnull().sum().sum()
print(f'Number of Missing Values: {missing_values}')

Number of Missing Values: 0


In [40]:
types = df.dtypes.value_counts()
print('Data Types and Frequency in Dataset:')
print(types)

Data Types and Frequency in Dataset:
float64    6
object     6
int64      4
Name: count, dtype: int64


In [41]:
catcols = df.select_dtypes(exclude = ['int64','float64']).columns
intcols = df.select_dtypes(include = ['int64']).columns
floatcols = df.select_dtypes(include = ['float64']).columns
# codificación
df = pd.get_dummies(df, columns = catcols,dtype=int,drop_first=True)
df

Rented Bike Count  Hour  Temperature(C)  Humidity(%)  Wind speed (m/s)  \
0                   254     0            -5.2           37               2.2   
1                   204     1            -5.5           38               0.8   
2                   173     2            -6.0           39               1.0   
3                   107     3            -6.2           40               0.9   
4                    78     4            -6.0           36               2.3   
...                 ...   ...             ...          ...               ...   
8755               1003    19             4.2           34               2.6   
8756                764    20             3.4           37               2.3   
8757                694    21             2.6           39               0.3   
8758                712    22             2.1           41               1.0   
8759                584    23             1.9           43               1.3   

      Visibility (10m)  Dew point temperature(C)  Solar Radiation (MJ/m2)  \
0                 2000                     -17.6                      0.0   
1                 2000                     -17.6                      0.0   
2                 2000                     -17.7                      0.0   
3                 2000                     -17.6                      0.0   
4                 2000                     -18.6                      0.0   
...                ...                       ...                      ...   
8755              1894                     -10.3                      0.0   
8756              2000                      -9.9                      0.0   
8757              1968                      -9.9                      0.0   
8758              1859                      -9.8                      0.0   
8759              1909                      -9.3                      0.0   

      Rainfall(mm)  Snowfall (cm)  Seasons_Spring  Seasons_Summer  \
0              0.0            0.0               0               0   
1              0.0            0.0               0               0   
2              0.0            0.0               0               0   
3              0.0            0.0               0               0   
4              0.0            0.0               0               0   
...            ...            ...             ...             ...   
8755           0.0            0.0               0               0   
8756           0.0            0.0               0               0   
8757           0.0            0.0               0               0   
8758           0.0            0.0               0               0   
8759           0.0            0.0               0               0   

      Seasons_Winter  Holiday_No Holiday  Functioning Day_Yes  \
0                  1                   1                    1   
1                  1                   1                    1   
2                  1                   1                    1   
3                  1                   1                    1   
4                  1                   1                    1   
...              ...                 ...                  ...   
8755               0                   1                    1   
8756               0                   1                    1   
8757               0                   1                    1   
8758               0                   1                    1   
8759               0                   1                    1   

      Dia Semana_jueves  Dia Semana_lunes  Dia Semana_martes  \
0                     0                 0                  0   
1                     0                 0                  0   
2                     0                 0                  0   
3                     0                 0                  0   
4                     0                 0                  0   
...                 ...               ...                ...   
8755                  0                 0                  0  

In [42]:
print('New Number of Features: %d'%(df.shape[1]))

New Number of Features: 35


In [43]:
pd.set_option('display.max_columns', None)  # Para mostrar todas las columnas
df

Rented Bike Count  Hour  Temperature(C)  Humidity(%)  Wind speed (m/s)  \
0                   254     0            -5.2           37               2.2   
1                   204     1            -5.5           38               0.8   
2                   173     2            -6.0           39               1.0   
3                   107     3            -6.2           40               0.9   
4                    78     4            -6.0           36               2.3   
...                 ...   ...             ...          ...               ...   
8755               1003    19             4.2           34               2.6   
8756                764    20             3.4           37               2.3   
8757                694    21             2.6           39               0.3   
8758                712    22             2.1           41               1.0   
8759                584    23             1.9           43               1.3   

      Visibility (10m)  Dew point temperature(C)  Solar Radiation (MJ/m2)  \
0                 2000                     -17.6                      0.0   
1                 2000                     -17.6                      0.0   
2                 2000                     -17.7                      0.0   
3                 2000                     -17.6                      0.0   
4                 2000                     -18.6                      0.0   
...                ...                       ...                      ...   
8755              1894                     -10.3                      0.0   
8756              2000                      -9.9                      0.0   
8757              1968                      -9.9                      0.0   
8758              1859                      -9.8                      0.0   
8759              1909                      -9.3                      0.0   

      Rainfall(mm)  Snowfall (cm)  Seasons_Spring  Seasons_Summer  \
0              0.0            0.0               0               0   
1              0.0            0.0               0               0   
2              0.0            0.0               0               0   
3              0.0            0.0               0               0   
4              0.0            0.0               0               0   
...            ...            ...             ...             ...   
8755           0.0            0.0               0               0   
8756           0.0            0.0               0               0   
8757           0.0            0.0               0               0   
8758           0.0            0.0               0               0   
8759           0.0            0.0               0               0   

      Seasons_Winter  Holiday_No Holiday  Functioning Day_Yes  \
0                  1                   1                    1   
1                  1                   1                    1   
2                  1                   1                    1   
3                  1                   1                    1   
4                  1                   1                    1   
...              ...                 ...                  ...   
8755               0                   1                    1   
8756               0                   1                    1   
8757               0                   1                    1   
8758               0                   1                    1   
8759               0                   1                    1   

      Dia Semana_jueves  Dia Semana_lunes  Dia Semana_martes  \
0                     0                 0                  0   
1                     0                 0                  0   
2                     0                 0                  0   
3                     0                 0                  0   
4                     0                 0                  0   
...                 ...               ...                ...   
8755                  0                 0                  0  

In [45]:
types = df.dtypes.value_counts()
print('Data Types and Frequency in Dataset:')
print(types)

Data Types and Frequency in Dataset:
int64      29
float64     6
Name: count, dtype: int64


In [46]:
X= df.drop(columns='Rented Bike Count')

In [47]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn import metrics
import numpy as np

# Separar variables predictoras (X) y la variable objetivo (y)
y = df['Rented Bike Count']

# Dividir los datos en conjuntos de entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Crear y entrenar el modelo
model = LinearRegression()
model.fit(X_train, y_train)

# Realizar predicciones
y_pred = model.predict(X_test)


In [48]:
import numpy as np

# mean absolute error
MAE = metrics.mean_absolute_error(y_test, y_pred)

# mean squared error
MSE = metrics.mean_squared_error(y_test, y_pred)

# root mean squared error
RMSE = np.sqrt(MSE)

In [49]:
print("MAE: ", MAE)
print("MSE: ", MSE)
print("RMSE: ", RMSE)

MAE:  301.9635144409125
MSE:  157361.03941644565
RMSE:  396.6875841470787


Variables con p-valores mayores a 0.05:
Dew point temperature(C) (p = 0.074)
Seasons_Summer (p = 0.437)
Seasons_Winter (p = 0.180)
Dia Semana_sábado (p = 0.213)
Mes_abril (p = 0.997)
Mes_enero (p = 0.447)
Mes_septiembre (p = 0.355)

In [50]:
from sklearn.model_selection import cross_val_score

# usar MSE - error cuadrático medio
scores = cross_val_score(model, X, y, cv=10, scoring='neg_mean_squared_error')
mse_scores = - scores
#print(mse_scores)

In [51]:
scores = cross_val_score(model, X, y, cv=10, scoring='neg_mean_absolute_error')

mae_scores = -scores

#print("MAE scores for each fold:", mae_scores)
print("Mean MAE:", np.mean(mae_scores))

Mean MAE: 323.5208318716987


In [55]:
rmse_scores = np.sqrt(mse_scores)

In [53]:
print(rmse_scores.mean())

416.3598047256507


In [56]:
print(mse_scores.mean())

184258.91437518768


In [57]:
import statsmodels.api as sm

y = df['Rented Bike Count']

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

# agregar constante explíticamente
X_train = sm.add_constant(X_train)

# regresión usando mínimos cuadrados ordinarios (ordinary least squares - OLS)
md = sm.OLS(y_train, X_train).fit()

# resumen de resultados
print(md.summary())

                            OLS Regression Results                            
Dep. Variable:      Rented Bike Count   R-squared:                       0.632
Model:                            OLS   Adj. R-squared:                  0.630
Method:                 Least Squares   F-statistic:                     361.4
Date:                Thu, 05 Sep 2024   Prob (F-statistic):               0.00
Time:                        13:50:54   Log-Likelihood:                -48558.
No. Observations:                6570   AIC:                         9.718e+04
Df Residuals:                    6538   BIC:                         9.740e+04
Df Model:                          31                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
const                   